text-generation-inference/server/text_generation_server/models/causal_lm.py

```diff
@@ -202,37 +202,51 @@ def filter(self, request_ids: List[int]) -> Optional["CausalLMBatch"]:
         # Apply indices to input_ids, attention mask, past key values and other items that need to be cached
         input_ids = self.input_ids[keep_indices]
         position_ids = self.position_ids[keep_indices]
-        self.attention_mask = self.attention_mask[
-            keep_indices,
-            -(self.padding_right_offset + max_input_length) : (
-                self.attention_mask.shape[1] - self.padding_right_offset
-            )
-            + new_padding_right_offset,
-        ]
+        if is_optimized_for_gaudi:
+            self.attention_mask = self.attention_mask[keep_indices]
+        else:
+            self.attention_mask = self.attention_mask[
+                keep_indices,
+                -(self.padding_right_offset + max_input_length) : (
+                    self.attention_mask.shape[1] - self.padding_right_offset
+                )
+                + new_padding_right_offset,
+            ]
 
         # Ensure that past_key_values tensors can be updated in-place
+        kv_tuple = False
```

Comment:
Why is this necessary now? Are layers automatically converted into lists at some point?

Reply:
in https://github.com/huggingface/optimum-habana/pull/389/files#diff-244871853752a7a019853c852803a797a9a14dddb3d9d859ecbae4b316ed6974R247. the past_key_value type should be exactly same with the case in single batch. or else hpu graph compile will meet type check error.